### PR TITLE
Do not use inline scripts

### LINKED ISSUE #145: Inline scripts break Content Security Policy
For security reasons, it's a good idea to have a [Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) saying something like `script-src: 'self'` - only scripts from this website allowed, and no inline scripts. This offers a good protection layer against Cross-Site Scripting attacks.

Now this doesn't quite work with MemE, two scripts are potentially defined inline: theme switcher and HTTPS switcher. It seems that the possible solutions are:

* Disable both features - possible but I actually liked the theme switcher.
* Allow these scripts in CSP based on their SHA-256 hash - possible but complicated, might also break when the theme is updated (slight changes to the script).
* Move the scripts for these features into a separate script file - adds one more request for a tiny file.
* Merge these scripts into the existing script file and include it at the top of the page - will delay page load minimally but is my preferred solution so far.

What do you think?

### PR DESCRIPTION
**Pull request by palant**
The first commit here moves loading JavaScript to the top of the page, which requires putting lots of code into a DOMContentLoaded event listener - this code originally expected all the elements to be present already. The theme switcher's inline script has been removed (fixes #145 for me).

I will add more commits to get rid of other inline scripts as well.

### CODE REVIEW

**Comment by palant** [comment]
Extensive changes to `script.html` in the latest commit:
* Not using `.Scratch` unnecessarily any more.
* Scripts are processed only when needed rather than always at the top of the template.
* Moved the generic part of script processing into a template to avoid repeating the code.

Note that I renamed `$Deliver` into `$` for this template - it should be more obvious this way. What do you think, maybe do it similarly in other templates as well?

**Comment by reuixiy** [comment]
> What do you think, maybe do it similarly in other templates as well?

Wow! That would be great!

**Comment by palant** [comment]
> Not using `.Scratch` unnecessarily any more.

I messed things up here. Scratch is actually necessary to get scripts from other templates, e.g. `lunr-search.html`. But maybe we should use a private scratch rather than a global one...

**Comment by palant** [comment]
This is a lot more messy than I expected, with those third-party scripts expecting to be loaded at a particular time...

**Comment by palant** [comment]
I guess I'll keep the various third-party scripts unchanged at the bottom and only fix the built-in functionality.

**Comment by palant** [comment]
Third-party scripts are now in a separate template, back at the bottom of the page - no functionality change here. This allowed making the logic for the built-in scripts simpler again.

File: layouts/partials/script.html [hunk 1]
```diff
@@ -1,106 +1,61 @@
...
-{{ partial "third-party/busuanzi.html" . }}
+{{- if .Site.Params.enableLunrSearch -}}
+    {{- $scripts = union $scripts (partial "third-party/lunr-search.html" .) -}}
+{{- end -}}
 
-{{ partial "custom/script.html" . }}
+{{- $scripts = union $scripts (slice "js/custom.js") -}}
 
-{{- .Scratch.Add "script" (slice $custom) -}}
+{{- $processedScripts := slice ("" | resources.FromString "dummy.js") -}}
```
Comment (palant) [inline layouts/partials/script.html:43]: This is a weird but it cannot be an empty slice. Apparently, slices are typed and an empty slice will be created with the wrong type making `resources.Concat` fail.

**Comment by palant** [comment]
Somehow I only noticed a bug after uploading this to a real website, it wasn't visible in local testing: `getCurrentTheme()` would error out because styles didn't load yet and the property it attempted to parse didn't exist. I pushed https://github.com/reuixiy/hugo-theme-meme/commit/e08e25dd29e8e807dba2681d2fc7e4e3c37665b7 to fix this.

**Pull request merged**